stash.c

```diff
@@ -0,0 +1,74 @@
+#include "stash.h"
+int stash_non_patch(const char* tmp_indexfile, const char* tmp_stagefile,
+                    const char* i_tree, const char* prefix)
+{
+	int iReturn;
+	char* szIndexFile = (char*)malloc(strlen(tmp_indexfile)+18);
+	sprintf(szIndexFile,"--index-output=\"%s\"",tmp_indexfile);
+	int iStageFile = open(tmp_stagefile, O_RDWR);
+	struct child_process read_tree = CHILD_PROCESS_INIT;
+	int iLen;
+	//strlen("read-tree") == 9
+	if(strlen(szIndexFile)>=strlen(i_tree)&&strlen(i_tree)>=9)
+        iLen=strlen(szIndexFile)+1;
+    else if(strlen(i_tree)>=strlen(szIndexFile)&&strlen(szIndexFile)>=9)
+        iLen=strlen(i_tree)+1;
+    else
+        iLen=10;
+	const char** read_tree_argv = (const char**)malloc(iLen*5);
+        *read_tree_argv="read-tree";
+        *(read_tree_argv+iLen)=szIndexFile;
+        *(read_tree_argv+iLen*2)="-m";
+        *(read_tree_argv+iLen*3)=i_tree;
+        *(read_tree_argv+iLen*4)=NULL;
+	read_tree.argv =read_tree_argv;
+	struct child_process diff = CHILD_PROCESS_INIT;
+	//strlen("--name-only") == 11
+	const char** diff_argv = (const char**)malloc(72);
+	    *diff_argv = "diff";
+	    *(diff_argv+12) = "-name-only";
+	    *(diff_argv+24) = "-z";
+	    *(diff_argv+36) = "HEAD";
+	    *(diff_argv+48) = "--";
+        *(diff_argv+60) = NULL;
+	diff.argv =diff_argv;
+	struct child_process update_index = CHILD_PROCESS_INIT;
+	//strlen("update-index") == 12
+	const char** update_index_argv = (const char**)malloc(78);
+	    *update_index_argv = "update-index";
+	    *(update_index_argv+13) = "-z";
+	    *(update_index_argv+26) = "--add";
+	    *(update_index_argv+39) = "--remove";
+	    *(update_index_argv+52) = "--stdin";
+        *(update_index_argv+65) = NULL;
+	update_index.argv =update_index_argv;
+	struct child_process write_tree = CHILD_PROCESS_INIT;
+	//strlen("write-tree") == 10
+	const char** write_tree_argv = (const char**)malloc(22);

```

Comment:
>That's because malloc() measures the size in bytes. So, we need to multiply the number of elements by the size of each element

Yes, you are right, but I initially waynted to allocate space `iLen * 5` `char` elements. Those are defined to always have a size of 1 byte.

>Furthermore, the run_command API expects the argv to be a const char** means that each element in the allocated array buffer is a pointer to a separate string, so we do not need to allocate memory for the string in the array itself. In other words,
const char** read_tree_argv = (const char**)malloc(5*sizeof(char*));
*read_tree_argv="read-tree";
*(read_tree_argv+1)=szIndexFile;
*(read_tree_argv+2)="-m";
*(read_tree_argv+3)=i_tree;
*(read_tree_argv+4)=NULL;
read_tree.argv =read_tree_argv;
In fact, since the size of the array is known, we can just skip the malloc()s and allocate the array on the stack directly:
const char* read_tree_argv[] = {"read-tree", szIndexFile, "-m", i_tree, NULL};

I went with 
````C

	const char* read_tree_argv[5];
        read_tree_argv[0] = "read-tree";
        read_tree_argv[1] = szIndexFile;
        read_tree_argv[2] = "-m";
        read_tree_argv[3] = i_tree;
        read_tree_argv[4] = NULL;
	read_tree.argv =read_tree_argv;
````
but your solution should work fine, too. 

>Hope this helps.

Should definitly help when my C becomes this rusty again. Thank you for your input.